src/core/bst.h

```diff
@@ -188,361 +133,540 @@ class BST {
     BST(BST_TYPE::BST_TYPE type) : type(type), lineno(0) {}
 #endif
     BST(BST_TYPE::BST_TYPE type, uint32_t lineno) : type(type), lineno(lineno) {}
-};
-
-class BST_expr : public BST {
-public:
-    virtual void* accept_expr(ExprVisitor* v) = 0;
 
-    BST_expr(BST_TYPE::BST_TYPE type) : BST(type) {}
-    BST_expr(BST_TYPE::BST_TYPE type, uint32_t lineno) : BST(type, lineno) {}
+    virtual bool isBST_dst() const { return false; }
 };
 
 class BST_stmt : public BST {
 public:
     virtual void accept_stmt(StmtVisitor* v) = 0;
 
-    int cxx_exception_count = 0;
-
     BST_stmt(BST_TYPE::BST_TYPE type) : BST(type) {}
+    BST_stmt(BST_TYPE::BST_TYPE type, uint32_t lineno) : BST(type, lineno) {}
 };
 
-class BST_slice : public BST {
+// base class of all nodes which have a single destination vreg
+class BST_dst : public BST_stmt {
```

Comment:
Only because you asked for thoughts on naming :P  When I first saw `BST_dst` I thought it was the destination that a bst node assigns to, not a bst node with a destination.  I have a very small preference for something like `BST_stmt_withdest`, and maybe calling `isBST_dst` something like `has_dest`.